contracts/nft.js

```diff
@@ -947,16 +947,23 @@ actions.setProperties = async (payload) => {
           // eslint-disable-next-line no-restricted-syntax
           for (const [name, data] of Object.entries(properties)) {
             const propertySchema = nft.properties[name];
+            let oldProperty = null;
             if (propertySchema.isReadOnly) {
               // read-only properties can only be set once
               if (api.assert(!(name in nftInstance.properties), 'cannot edit read-only properties')) {
                 nftInstance.properties[name] = data;
                 shouldUpdate = true;
               }
             } else {
+              oldProperty = nftInstance.properties[name];
               nftInstance.properties[name] = data;
               shouldUpdate = true;
             }
+            if (shouldUpdate) {
```

Comment:
One optimization here would be to call the handleNftSetProperty action only if the data property value has actually changed.  If the new value is same as old value, no need to call this.  It's a small thing, but even loading & calling the other contract has some processing cost and saving every little bit helps.